Bottom: 6d1e5bda9cc14bfa4c80b381432a064be7336f1e
Top:    9af18789fc4a504550058f002e75fbf7858e80e2
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-05-19 18:13:43 +0100

buffer.c: Update polygon r-tree when adding a polygon to the buffer.

This resulted in a crash when rotating a buffer containing a polygon,
as the polygon r-tree associated with the buffer was NULL despite the
polygon count being non-zero.

Reported-by: Gabriel Paubert <paubert@iram.es>


---

diff --git a/src/buffer.c b/src/buffer.c
index b396918..2e63af5 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -196,6 +196,15 @@ AddPolygonToBuffer (LayerTypePtr Layer, PolygonTypePtr Polygon)
 
   polygon = CreateNewPolygon (layer, Polygon->Flags);
   CopyPolygonLowLevel (polygon, Polygon);
+
+  /* Update the polygon r-tree. Unlike similarly named functions for
+   * other objects, CreateNewPolygon does not do this as it creates a
+   * skeleton polygon object, which won't have correct bounds.
+   */
+  if (!layer->polygon_tree)
+    layer->polygon_tree = r_create_tree (NULL, 0, 0);
+  r_insert_entry (layer->polygon_tree, (BoxType *)polygon, 0);
+
   CLEAR_FLAG (FOUNDFLAG | ExtraFlag, polygon);
   return (polygon);
 }
